### DISCUSSION TITLE
nginx proxy to Theia in docker container: configuring HTTP Host

### DISCUSSION THREAD

**Discussion by profbbrown** [eclipse-theia/theia#9912] [Q&A]
I have Theia running in Docker containers behind an nginx reverse proxy. My question pertains to the mini-browser. The URLs look like this:

    {UUID}.mini-browser.{host}/path

For example: `https://0b45d02d-7899-4587-9c52-79fe7609c451.mini-browser.theia.mydomain.com/mini-browser/home/image/myimage.png`

In order to pass the HTTP request in to Theia running in the container, I need to rewrite the hostname and rewrite the HTTP Host header.

I think I have the hostname rewrite rules working, since the rest of the IDE is functional. It's only the mini-browser that isn't working. What should the Host header look like?

**Comment by profbbrown** [comment]
I will answer my own question.

I used a regular expression for the server name to capture the unique mini-browser hostname. Inside the `server` directive for nginx:

    server_name ~^(?<mini_browser_host>[a-f0-9-]+\.mini-browser\.)?theia\.mydomain\.com$;

In the proxy directives in the `location` section, I used the `mini_browser_host` to set the HTTP `Host` header:

    ...
    proxy_set_header Host "${mini_browser_host}$http_host";
    ...

**Reply by Gopikrishna-GitHB** [reply]
@profbbrown, I am too facing the similar issue. Tried to use your solution but it didnt worked. Could you help on this. 
I am sharing my issue link below. If there is any further clarification, Please let me know. Thanks in advance.
https://github.com/eclipse-theia/theia/discussions/12044

**Reply by profbbrown** [reply]
Update on this issue:

I recently updated all my Theia instances to 1.46.0 and the mini-browser stopped working. Looking through these discussions it appears the new solution is to add an environment variable: `THEIA_WEBVIEW_EXTERNAL_ENDPOINT={{hostname}}`

I did this in the script that starts the container and it works fine, except for a warning that appears when the UI starts up that doing so is potentially a security vulnerability.

**Reply by msujew** [reply]
You can disable the warning (note that it is indeed a security issue to set the endpoint to just `{{hostname}}`) by updating your package.json, see [here](https://github.com/eclipse-theia/theia/tree/master/packages/plugin-ext#security-warnings).

**Reply by profbbrown** [reply]
Thats what I've done to make things functional for now. I'm interested in getting it to work "correctly," though.

**Reply by bhawanishiv-techolution** [reply]
Hi @profbbrown,
just needed your help to understand how you were able to run theia in docker container with the env: `THEIA_WEBVIEW_EXTERNAL_ENDPOINT={{hostname}}`?

I tried adding this in docker file but theia seems not load read the env even from the container.


Here is my docker file contents:


```
# Builder stage
FROM node:18-bullseye as build-stage


# install required tools to build the application
RUN apt-get update && apt-get install -y libxkbfile-dev libsecret-1-dev

WORKDIR /home/theia


# Copy repository files
COPY . .


ENV THEIA_WEBVIEW_ENDPOINT_PATTERN='{{hostname}}'
# TODO: uncomment this if build fails
ENV PUPPETEER_SKIP_DOWNLOAD=false

# Remove unnecesarry files for the browser application
# Download plugins and build application production mode
# Use yarn autoclean to remove unnecessary files from package dependencies
RUN yarn global add node-gyp && \
    yarn --skip-integrity-check --network-timeout 100000

RUN yarn -s download:plugins && \
    yarn browser build:production

RUN yarn --production && \
    yarn autoclean --init && \
    echo *.ts >> .yarnclean && \
    echo *.ts.map >> .yarnclean && \
    echo *.spec.* >> .yarnclean && \
    yarn autoclean --force && \
    yarn cache clean && \
    rm -r node_modules
# rm -r .git applications/electron theia-extensions/launcher theia-extensions/updater node_modules


# Production stage uses a small base image
FROM node:18-bullseye-slim as production-stage

# Create theia user and directories
# Application will be copied to /home/theia
# Default workspace is located at /home/project
RUN adduser --system --group theia
RUN chmod g+rw /home && \
    mkdir -p /home/project && \
    chown -R theia:theia /home/theia && \
    chown -R theia:theia /home/project;

# Install required tools for application: Temurin JDK, JDK, SSH, Bash, Maven, python, python3
# Node is already available in base image
RUN apt-get update && apt-get install -y wget apt-transport-https && \
    wget -O - https://packages.adoptium.net/artifactory/api/gpg/key/public | tee /usr/share/keyrings/adoptium.asc && \
    echo "deb [signed-by=/usr/share/keyrings/adoptium.asc] https://packages.adoptium.net/artifactory/deb $(awk -F= '/^VERSION_CODENAME/{print$2}' /etc/os-release) main" | tee /etc/apt/sources.list.d/adoptium.list && \
    apt-get update && apt-get install -y git openssh-client openssh-server bash libsecret-1-0 python python3 && \
    apt-get purge -y wget && \
    apt-get clean

ENV HOME /home/theia
WORKDIR /home/theia

# Copy application from builder-stage
COPY --from=build-stage --chown=theia:theia /home/theia /home/theia

EXPOSE 3000

ENV THEIA_WEBVIEW_ENDPOINT_PATTERN='{{hostname}}'

# Specify default shell for Theia and the Built-In plugins directory
ENV SHELL=/bin/bash \
    THEIA_DEFAULT_PLUGINS=local-dir:/home/theia/plugins

# Use installed git instead of dugite
ENV USE_LOCAL_GIT true

# Swtich to Theia user
USER theia
WORKDIR /home/theia

# Launch the backend application via node
ENTRYPOINT ["node", "/home/theia/examples/browser/lib/backend/main.js" ]

# Arguments passed to the application
# Change the environment to "host"
CMD [ "/home/project", "--hostname=0.0.0.0" ]
```

Just to inform, I have built the theia browser and running it via node {main_script_file.js} (which seems to be compiled by webpack).

Your kind support is much appreciated for me understand and resolve this.

**Reply by profbbrown** [reply]
The environment variable I used was `THEIA_WEBVIEW_EXTERNAL_ENDPOINT`

**Comment by profbbrown** [comment]
When I started this thread in 2021 I was running the then-current version of Theia, something like 1.25. In 2024 I have upgraded to 1.46 and the webview has stopped working. 

Here is a summary of my setup.

Theia is running in a Docker container. The container has exposed port 3000. An Nginx reverse proxy handles incoming requests and funnels them to the correct container.

An incoming URL normally looks like `https://theia.mydomain.com/ide/USERNAME/WORKSPACE/whatever`, where USERNAME and WORKSPACE are identifiers for a unique container. The reverse proxy turns this into `http://theia-USERNAME-WORKSPACE:3000/whatever` and uses the Docker resolver to send the request on to the correct container.

For the old mini-browser, I had a regular expression server_name match to pull out the UUID. It was documented in my first reply, but I'll include it here:

```server_name ~^(?<mini_browser_host>[a-f0-9-]+\.(mini-browser|webview)\.)?theia\.mydomain\.com$;```

Later there is a `location` section that performs the reverse proxy. In particular, it rewrites the hostname by setting the `proxy_host_port` variable, rewrites the URL path portion by stripping off the `/ide/USERNAME/WORKSPACE` prefix, and sets the `Host` header by including the `mini_browser_host` variable.

```
location ~ ^/ide/(?<username>(\w+))/(?<wksp>(\w+)) {
    resolver 127.0.0.11;    # Docker resolver
    set $proxy_host_port "theia-$username-$wksp:3000";
    rewrite ^/ide/\w+/\w+(/.*) $1 break;
    root   /var/www/html;
    auth_request /auth.php;
    proxy_http_version 1.1;
    proxy_buffering off;
    proxy_set_header Host "${mini_browser_host}$http_host";
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $proxy_x_forwarded_proto;
    proxy_set_header X-Forwarded-Ssl $proxy_x_forwarded_ssl;
    proxy_set_header X-Forwarded-Port $proxy_x_forwarded_port;
    proxy_set_header Proxy "";     # Mitigate httpoxy attack
    proxy_pass http://$proxy_host_port;
  }
```

In Theia 1.46, this setup no longer works. A JavaScript file is failing to load. Inspection of the web page shows that the failing URL is `https://d517905a-21a9-5591-9981-2db6e3fd32d9.mini-browser.theia.mydomain.com/ide/USERNAME/WORKSPACE/webview/theia-resource/file///theia/plugins/vscode.media-preview/extension/media/imagePreview.js`

It looks like there is a `file://` URL embedded in there. Is that correct?

**Reply by msujew** [reply]
> It looks like there is a file:// URL embedded in there. Is that correct?

Yes, the part after `theia-resource` represents the URI of the loaded file, which might not necessarily be of scheme `file`. Consequently having `file` in there is required to correctly locate the expected resource.

**Reply by profbbrown** [reply]
Thanks for the response. 

If I set the endpoint to `{{hostname}}` the webview works fine. But when set back to the default, it doesn't work. The only difference is the hostname that includes the UUID. So it's probably something in my Ngnix configuration.

What `Host` header is the backend expecting when the user requests a document to be rendered with the webview? For example, if the incoming host is `d517905a-21a9-5591-9981-2db6e3fd32d9.mini-browser.theia.mydomain.com`, what should it look like to the backend that is running in a container?

**Reply by msujew** [reply]
Theia serves the subdomain as `{{guid}}.mini-browser.{{hostname}}`/``{{guid}}.webview.{{hostname}}`` (where `{{hostname}}` is usually something like `localhost:3000`). So you need to rewrite it to `d517905a-21a9-5591-9981-2db6e3fd32d9.mini-browser.localhost:3000`.
